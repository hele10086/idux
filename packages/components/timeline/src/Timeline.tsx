/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { TimelinePlacement, TimelineProps } from './types'
import type { VNode } from 'vue'

import { cloneVNode, defineComponent, h } from 'vue'

import { IxIcon } from '@idux/components/icon'

import IxTimelineItem from './TimelineItem'
import { timelineProps } from './types'

const timelinePrefixCls = 'ix-timeline'
const itemPrefixCls = 'ix-timeline-item'

export default defineComponent({
  name: 'IxTimeline',
  props: timelineProps,
  setup(props, { slots }) {
    return () => {
      const pendingSlots = slots?.pending?.()
      const pendingDotSlots = slots?.pendingDot?.()
      const defaultSlots = slots?.default?.() || []
      const pendingNode = pendingSlots || props.pending
      const pendingDotNode = pendingDotSlots || props.pendingDot || h(IxIcon, { name: 'loading' })
      let penddingItem: null | VNode = null

      if (pendingNode) {
        penddingItem = h(
          IxTimelineItem,
          { class: `${itemPrefixCls}-pending-dot` },
          { default: () => pendingNode, dot: () => pendingDotNode },
        )
      }

      const timelineItems = props.reverse ? [penddingItem, ...defaultSlots.reverse()] : [...defaultSlots, penddingItem]
      const nonNullItems = timelineItems.filter(item => !!item) as VNode[]
      const itemsLength = nonNullItems.length
      const items: VNode[] = []
      const itemPlacementArr: TimelinePlacement[] = []

      nonNullItems.forEach((item, index) => {
        const placement = getItemPlacement(item, index, props.placement)

        itemPlacementArr.push(placement)
        items.push(
          cloneVNode(item, {
            class: useItemClasses({
              hasPendingNode: !!pendingNode,
              placement,
              index,
              itemsLength,
              props: props,
            }),
          }),
        )
      })

      return h(
        'ul',
        {
          class: useClasses(itemPlacementArr, props),
        },
        items,
      )
    }
  },
})

const getItemPlacement = (node: VNode, index: number, placement: TimelinePlacement): TimelinePlacement => {
  const itemPlacement = node.props?.placement
  if (itemPlacement) {
    return itemPlacement
  }

  if (placement === 'alternate') {
    return index % 2 ? 'start' : 'end'
  }

  return placement
}

const getRealPlacement = (itemPlacementArr: TimelinePlacement[]): TimelinePlacement => {
  return itemPlacementArr.reduce((result: TimelinePlacement, nextPlacement, index) => {
    if (!index) {
      result = nextPlacement
    } else if (result !== nextPlacement) {
      result = 'alternate'
    }

    return result
  }, 'end')
}

const useClasses = (itemPlacementArr: TimelinePlacement[], props: TimelineProps): string => {
  const realPlacement = getRealPlacement(itemPlacementArr)
  const placementCls = `${timelinePrefixCls}-${realPlacement}`

  let cls = `${timelinePrefixCls} ${placementCls}`

  if (props.reverse) {
    cls += ` ${timelinePrefixCls}-reverse`
  }

  if (!props.both) {
    cls += ` ${timelinePrefixCls}-not-both`
  }

  return cls
}

const useItemClasses = ({
  hasPendingNode,
  placement,
  index,
  itemsLength,
  props,
}: {
  hasPendingNode: boolean
  placement: TimelinePlacement
  index: number
  itemsLength: number
  props: TimelineProps
}): string => {
  let cls = `${itemPrefixCls}-${placement}`

  if (hasPendingNode) {
    const isReversePending = props.reverse && index === 0
    const isPending = !props.reverse && index === itemsLength - 2

    if (isReversePending || isPending) {
      cls += ` ${itemPrefixCls}-pending`
    }
  }

  return cls
}
