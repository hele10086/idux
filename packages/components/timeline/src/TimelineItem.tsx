/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { CSSProperties } from 'vue'

import { computed, defineComponent } from 'vue'

import { hasSlot } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/config'
import { isPresetColor, isStatusColor } from '@idux/components/utils'

import { timelineItemProps } from './types'

export default defineComponent({
  name: 'IxTimelineItem',
  props: timelineItemProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-timeline`)
    const prefixCls = mergedPrefixCls.value
    const isPresetOrStatus = computed(() => isPresetColor(props.color) || isStatusColor(props.color))

    const dotStyle = computed(() => {
      if (isPresetOrStatus.value) {
        return {}
      }
      return {
        color: props.color,
        'border-color': props.color,
      }
    })

    const dotClass = computed(() => {
      return {
        [`${prefixCls}-item-dot`]: true,
        [`${prefixCls}-item-dot-custom`]: hasSlot(slots, 'dot') || !!props.dot,
        [`${prefixCls}-item-dot-${props.color}`]: isPresetOrStatus.value,
      }
    })

    return () => {
      return (
        <li class={`${prefixCls}-item`}>
          <div class={`${prefixCls}-item-line`}></div>
          <div class={dotClass.value} style={dotStyle.value as CSSProperties}>
            {slots.dot ? slots.dot() : props.dot}
          </div>
          <div class={`${prefixCls}-item-content`}>
            <div class={`${prefixCls}-item-label`}>{props.label || slots.label?.()}</div>
            <div class={`${prefixCls}-item-desc`}>{slots.default?.()}</div>
          </div>
        </li>
      )
    }
  },
})
