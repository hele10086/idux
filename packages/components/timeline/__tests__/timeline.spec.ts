import { MountingOptions, mount } from '@vue/test-utils'
import { defineComponent, h } from 'vue'

import { renderWork } from '@tests'

import IxTimeline from '../src/Timeline'
import IxTimelineItem from '../src/TimelineItem'
import { TimelineProps } from '../src/types'

const TimelineDefaultSlots = defineComponent({
  components: { IxTimelineItem },
  render() {
    return [
      h(IxTimelineItem, {}, { default: () => 0 }),
      h(IxTimelineItem, {}, { default: () => 1 }),
      h(IxTimelineItem, {}, { default: () => 2 }),
      h(IxTimelineItem, {}, { default: () => 3 }),
    ]
  },
})

describe('Timeline', () => {
  const TimelineMount = (options?: MountingOptions<Partial<TimelineProps>>) => mount(IxTimeline, { ...options })

  renderWork(IxTimeline)

  test('placement work', async () => {
    const wrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    expect(wrapper.findAll('.ix-timeline-item').length).toBe(4)
    expect(wrapper.findAll('.ix-timeline-item-end').length).toBe(4)
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ placement: 'start' })
    expect(wrapper.findAll('.ix-timeline-item-start').length).toBe(4)
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ placement: 'alternate' })
    expect(wrapper.findAll('.ix-timeline-item-start').length).toBe(2)
    expect(wrapper.findAll('.ix-timeline-item-end').length).toBe(2)
    wrapper.findAll('.ix-timeline-item').forEach((item, index) => {
      const placement = index % 2 ? 'start' : 'end'

      expect(item.classes()).toContain(`ix-timeline-item-${placement}`)
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('reverse work', async () => {
    const wrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
      },
    })
    wrapper.findAll('.ix-timeline-item').forEach((item, index) => {
      expect(item.text()).toBe(index.toString())
    })
    expect(wrapper.html()).toMatchSnapshot()

    const reverseWrapper = TimelineMount({
      props: {
        reverse: true,
      },
    })
    reverseWrapper.findAll('.ix-timeline-item').forEach((item, index) => {
      expect(item.text()).toBe((3 - index).toString())
    })
    expect(reverseWrapper.html()).toMatchSnapshot()
  })

  test('pending work', async () => {
    const wrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
      },
    })
    const pendingText = 'pendingText'

    expect(wrapper.findAll('.ix-timeline-item').length).toBe(4)
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ pending: true })
    expect(wrapper.findAll('.ix-timeline-item').length).toBe(5)
    expect(wrapper.findAll('.ix-timeline-item')[3].classes()).toContain('ix-timeline-item-pending')
    expect(wrapper.findAll('.ix-timeline-item')[4].classes()).toContain('ix-timeline-item-pending-dot')
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ pending: pendingText })
    expect(wrapper.findAll('.ix-timeline-item').length).toBe(5)
    expect(wrapper.findAll('.ix-timeline-item')[3].classes()).toContain('ix-timeline-item-pending')
    expect(wrapper.findAll('.ix-timeline-item')[4].classes()).toContain('ix-timeline-item-pending-dot')
    expect(wrapper.findAll('.ix-timeline-item')[4].text()).toBe(pendingText)
    expect(wrapper.html()).toMatchSnapshot()

    const reverseWrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
      },
      props: {
        pending: true,
        reverse: true,
      },
    })

    reverseWrapper.findAll('.ix-timeline-item').forEach((item, index) => {
      if (index === 0) {
        expect(item.classes()).toContain('ix-timeline-item-pending')
        expect(item.classes()).toContain('ix-timeline-item-pending-dot')
      } else {
        expect(item.text()).toBe((4 - index).toString())
      }
    })
    expect(reverseWrapper.html()).toMatchSnapshot()
  })

  test('pending slot work', async () => {
    const pendingSlotText = 'pendingSlotText'
    const pendingText = 'pendingText'
    const wrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
        pending: pendingSlotText,
      },
    })

    expect(wrapper.findAll('.ix-timeline-item').length).toBe(5)
    expect(wrapper.findAll('.ix-timeline-item')[3].classes()).toContain('ix-timeline-item-pending')
    expect(wrapper.findAll('.ix-timeline-item')[4].classes()).toContain('ix-timeline-item-pending-dot')
    expect(wrapper.findAll('.ix-timeline-item')[4].text()).toBe(pendingSlotText)
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ pending: pendingText })
    expect(wrapper.findAll('.ix-timeline-item')[4].text()).toBe(pendingSlotText)
    expect(wrapper.html()).toMatchSnapshot()

    const reverseWrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
        pending: pendingSlotText,
      },
      props: {
        reverse: true,
      },
    })

    reverseWrapper.findAll('.ix-timeline-item').forEach((item, index) => {
      if (index === 0) {
        expect(item.classes()).toContain('ix-timeline-item-pending')
        expect(item.classes()).toContain('ix-timeline-item-pending-dot')
      } else {
        expect(item.text()).toBe((4 - index).toString())
      }
    })
    expect(reverseWrapper.html()).toMatchSnapshot()
  })

  test('pendingDot work', async () => {
    const wrapper = TimelineMount({
      slots: {
        default: TimelineDefaultSlots,
      },
      props: {
        pending: true,
      },
    })
    const pendingDotText = 'pendingDotText'

    expect(wrapper.findAll('.ix-timeline-item')[4].find('.ix-timeline-item-dot').find('.ix-icon').classes()).toContain(
      'ix-icon-loading',
    )
    expect(wrapper.html()).toMatchSnapshot()

    await wrapper.setProps({ pendingDot: pendingDotText })
    expect(wrapper.findAll('.ix-timeline-item')[4].find('.ix-timeline-item-dot').text()).toBe(pendingDotText)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
