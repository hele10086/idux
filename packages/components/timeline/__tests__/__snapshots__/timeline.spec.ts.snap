// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timeline pending work 1`] = `
"<ul class=\\"ix-timeline ix-timeline-end\\">
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">0</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">1</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">2</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">3</div>
    </div>
  </li>
</ul>"
`;

exports[`Timeline pendingDot work 1`] = `
"<ul class=\\"ix-timeline ix-timeline-end\\">
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">0</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">1</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">2</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">3</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item ix-timeline-item-pending-dot ix-timeline-item-end\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-custom ix-timeline-item-dot-primary\\"><i class=\\"ix-icon ix-icon-loading\\" role=\\"img\\" aria-label=\\"loading\\"></i></div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">
        <!---->
      </div>
    </div>
  </li>
</ul>"
`;

exports[`Timeline pendingDot work 2`] = `
"<ul class=\\"ix-timeline ix-timeline-end\\">
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">0</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">1</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">2</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">3</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item ix-timeline-item-pending-dot ix-timeline-item-end\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-custom ix-timeline-item-dot-primary\\">pendingDotText</div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">
        <!---->
      </div>
    </div>
  </li>
</ul>"
`;

exports[`Timeline placement work 1`] = `
"<ul class=\\"ix-timeline ix-timeline-end\\">
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">0</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">1</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">2</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">3</div>
    </div>
  </li>
</ul>"
`;

exports[`Timeline render work 1`] = `"<ul class=\\"ix-timeline ix-timeline-end\\"></ul>"`;

exports[`Timeline reverse work 1`] = `
"<ul class=\\"ix-timeline ix-timeline-end\\">
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">0</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">1</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">2</div>
    </div>
  </li>
  <li class=\\"ix-timeline-item\\">
    <div class=\\"ix-timeline-item-line\\"></div>
    <div class=\\"ix-timeline-item-dot ix-timeline-item-dot-primary\\">
      <!---->
    </div>
    <div class=\\"ix-timeline-item-content\\">
      <div class=\\"ix-timeline-item-label\\">
        <!---->
      </div>
      <div class=\\"ix-timeline-item-desc\\">3</div>
    </div>
  </li>
</ul>"
`;

exports[`Timeline reverse work 2`] = `"<ul class=\\"ix-timeline ix-timeline-end ix-timeline-reverse\\"></ul>"`;
